components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
 
     /// Adapter function to retrieve an installed package given an
     /// identifier, if it exists.
-    fn installed_package(&self, ident: &PackageIdent) -> Option<PackageInstall> {
-        PackageInstall::load(ident, Some(self.fs_root_path)).ok()
+    fn installed_package(&self, ident: &FullyQualifiedPackageIdent) -> Option<PackageInstall> {
+        PackageInstall::load(ident.as_ref(), Some(self.fs_root_path)).ok()
+    }
+
+    /// Checks for the latest installed package or cached artifact that matches a given package
+    /// identifier and returns a fully qualified package identifer if a match exists.
+    fn latest_installed_or_cached(
+        &self,
+        ident: &PackageIdent,
+    ) -> Result<FullyQualifiedPackageIdent> {
+        let latest_installed = self.latest_installed_ident(&ident);
+        let latest_cached = self.latest_cached_ident(&ident);
+        debug!(
+                "latest installed: {:?}, latest_cached: {:?}",
+                &latest_installed,
+                &latest_cached,
+            );
+        let latest = match (latest_installed, latest_cached) {
+            (Ok(pkg_install), Err(_)) => pkg_install,
+            (Err(_), Ok(pkg_artifact)) => pkg_artifact,
+            (Ok(pkg_install), Ok(pkg_artifact)) => {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
```

Comment:
I think that should be the same. The reason I'm saying "think" however is that it takes me a lot longer to reason through the compound conditional than to scan through the `match` arms. That's probably a personal style thing though--for me I tend to use `match`ing where possible because the compiler makes sure I don't miss any arms (it helped in an earlier version of this code where I was dealing with `Option`s and had to convert to `Result`s). Same for that inner `if`/`else`--I'm trying to get a fully balanced conditional so that I can see every condition has been dealt with.

Reply:
👍 on using match... it's a lot easier to reason through, and you can be sure that the compiler has your back.

Incidentally, you'll find this same pattern of matching against a tuple in Erlang code in place of complex if/else logic, and for the same reason. It takes a little getting used to at first, but I think it's ultimately clearer.